compiler/transformation/transform.rs

```diff
@@ -61,3 +62,53 @@ fn prune_redundant_roleplayer_deduplication(conjunction: &mut Conjunction, block
         }
     });
 }
+
+pub fn optimize_away_statically_unsatisfiable_conjunctions(
+    conjunction: &mut Conjunction,
+    block_annotations: &TypeAnnotations,
+) -> bool {
+    let mut must_optimise_away = false;
+    for nested in conjunction.nested_patterns_mut() {
+        match nested {
+            NestedPattern::Disjunction(disjunction) => {
+                let mut optimised_unsatisfiable_branch_ids = Vec::new();
+                for branch in disjunction.conjunctions_mut().iter_mut() {
+                    if optimize_away_statically_unsatisfiable_conjunctions(branch, block_annotations) {
+                        optimised_unsatisfiable_branch_ids.push(branch.scope_id())
+                    }
+                }
+                disjunction.optimise_away_failing_branches(optimised_unsatisfiable_branch_ids);
+                must_optimise_away = must_optimise_away || disjunction.conjunctions().is_empty();
+            }
+            NestedPattern::Negation(negation) => {
+                optimize_away_statically_unsatisfiable_conjunctions(negation.conjunction_mut(), block_annotations);
+            }
+            NestedPattern::Optional(optional) => {
+                optimize_away_statically_unsatisfiable_conjunctions(optional.conjunction_mut(), block_annotations);
+            }
+        }
+    }
+    let must_optimise_away = must_optimise_away
+        || conjunction.constraints().iter().any(|constraint| {
+            if let Some(constraint_annotation) = block_annotations.constraint_annotations_of(constraint.clone()) {
+                match constraint_annotation {
+                    ConstraintTypeAnnotations::LeftRight(lr) => {
+                        debug_assert!(lr.left_to_right().is_empty() == lr.right_to_left().is_empty());
+                        lr.left_to_right().is_empty()
+                    }
+                    ConstraintTypeAnnotations::Links(links) => {
+                        links.player_to_role.is_empty() || links.relation_to_role.is_empty()
+                    }
+                    ConstraintTypeAnnotations::IndexedRelation(_) => {
+                        unreachable!("This is called before IndexedRelations are inserted")
+                    }
+                }
+            } else {
+                false
+            }
+        });
+    if must_optimise_away {
+        conjunction.optimise_away();
+    }
+    must_optimise_away
```

Comment:
You're returning a boolean when your function implies you're doing a stateful operation on the conjunction - which you are! As a result, after your function runs you now have two sources of truth: the conjunction, and this boolean. This is at risk of going out of sync at some point.

A tighter API is to make the caller determine whether the transformation was applied themselves: here we do `conjunction.set_unsatisfiable()`, and then the caller, if they must know, can check `conjunction.is_unsatisifable()`.

This is also safer for the future, if there are _other_ ways to achieve unsatisifable conjunctions! If a conjunction arrives at this transformation unsatisifiable, it'll still work nicely and combine that information with what it computes here.

Reply:
Ganesh, and then I, call this making "tight" APIs (you can't misuse them).

Reply:
> This is at risk of going out of sync at some point.
Sure. I could also not mutate the inner conjunctions and just optimise away the outer-most unsatisfiable conjunction.